crates/wysiwyg/src/dom/find_extended_range.rs

```diff
@@ -0,0 +1,221 @@
+// Copyright 2022 The Matrix.org Foundation C.I.C.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::{DomHandle, UnicodeString};
+
+use super::nodes::dom_node::DomNodeKind;
+use super::{find_range, Dom, DomLocation, Range};
+
+impl<S> Dom<S>
+where
+    S: UnicodeString,
+{
+    /// Returns a range from given start and end, extended up to the next
+    /// structure node, list item or linebreak on each side.
+    pub fn find_extended_range(&self, start: usize, end: usize) -> Range {
+        let (s, e) = self.find_extended_selection(start, end);
+        find_range::find_range(self, s, e)
+    }
+
+    /// Returns a selection from given start and end, extended up to the next
+    /// structure node, list item or linebreak on each side.
+    pub fn find_extended_selection(
+        &self,
+        start: usize,
+        end: usize,
+    ) -> (usize, usize) {
+        let range = find_range::find_range(self, start, end);
+        let leaves: Vec<&DomLocation> = range
+            .leaves()
+            .filter(|loc| loc.position < range.end())
+            .collect();
+        if leaves.is_empty() {
+            return (start, end);
+        }
+
+        let first_leaf = leaves.first().unwrap();
+        let last_leaf = leaves.last().unwrap();
+        let extended_start = start - self.extended_offset_before(first_leaf);
+        let extended_end = end + self.extended_offset_after(last_leaf);
+
+        (extended_start, extended_end)
+    }
+
+    /// Returns the offset before a location until a structure
+    /// node, list item or linebreak is found.
+    fn extended_offset_before(&self, location: &DomLocation) -> usize {
+        if location.kind == DomNodeKind::LineBreak {
+            // Item at location is a linebreak, no need to iterate
+            return location.start_offset;
+        }
+
+        let mut iter = self.iter_from_handle(&location.node_handle).rev();
+        // Remove location and use offset instead
+        iter.next();
+        let mut offset = location.start_offset;
+
+        for node in iter {
+            if node.is_text_node() {
+                offset += node.text_len();
+            } else if node.is_line_break() {
+                offset += node.text_len();
+                break;
+            } else if node.is_block_node() || node.is_list_item() {
+                break;
+            }
+        }
```

Comment:
Maybe personal preference but I wonder if something like the following would be easier to read
```suggestion
        for node in iter {
            if node.is_text_node() || node.is_line_break() {
                offset += node.text_len();
            } 
            
            if node.is_line_break() || node.is_block_node() || node.is_list_item() {
                break;
            }
        }
```